src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -1,64 +1,105 @@
- * else happen.
- * Tried to fix this by changing the sampletime and clock prescaler values, but
- * nothing worked.
- * This was "fixed" by commenting out the NVIC_EnableIRQ, stopping the interrupt
- * from ever being enabled.
+ * f4xx DMA ADC uses a timer to trigger ADC conversions. Timer frequency is 1kHz.
+ *
+ * Timers were added to slow down ADC DMA interrupts, as when they ADC DMA is allowed to convert constantly,
+ * the interrupts stopped the program from doing anything else besides interrupt calls.
  *
- * For commit from before code was removed, refer to commit 81624521a8b2c4b66480193e88cf32782aaee84d.
+ * WARNING: DOES NOT WORK ON PINS THAT ARE ALREADY IN USE (LIKE UART PINS)
  */
 
 #include <HALf4/stm32f4xx.h>
 #include <HALf4/stm32f4xx_hal_adc.h>
 #include <core/io/pin.hpp>
 #include <core/io/platform/f4xx/ADCf4xx.hpp>
 #include <core/io/platform/f4xx/GPIOf4xx.hpp>
+#include <core/platform/f4xx/stm32f4xx.hpp>
+#include <core/utils/log.hpp>
+
+namespace {
+/// This is made as a global variable so that it is accessible in the interrupt.
+DMA_HandleTypeDef* dmaHandle[3];
+ADC_HandleTypeDef* adcHandle[3];
+} // namespace
+
+/**
+ * @brief This function handles DMA2 stream0 global interrupt. (For ADC 1)
+ */
+extern "C" void DMA2_Stream0_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[0]);
+    HAL_ADC_IRQHandler(adcHandle[0]);
+}
+
+/**
+ * @brief This function handles DMA2 stream2 global interrupt. (For ADC 2)
+ */
+extern "C" void DMA2_Stream2_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[1]);
+    HAL_ADC_IRQHandler(adcHandle[1]);
+}
+
+/**
+ * @brief This function handles DMA2 stream1 global interrupt. (For ADC 3)
+ */
+extern "C" void DMA2_Stream1_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[2]);
+    HAL_ADC_IRQHandler(adcHandle[2]);
+}
 
 namespace core::io {
+#define ADC1_SLOT 0
+#define ADC2_SLOT 1
+#define ADC3_SLOT 2
+
+#define ADC1SHIFT 5
+#define ADC2SHIFT 6
+#define ADC3SHIFT 7
 
-// Init static member variables
-ADC_HandleTypeDef ADCf4xx::halADC = {0};
-Pin ADCf4xx::channels[MAX_CHANNELS];
-uint16_t ADCf4xx::buffer[MAX_CHANNELS];
-DMA_HandleTypeDef ADCf4xx::halDMA = {0};
+// Combines the channel memory value with the ADC peripherals it supports into one uint32_t
+#define CHANNEL_SET(adc1, adc2, adc3, ch) (ch | (adc1 << ADC1SHIFT) | (adc2 << ADC2SHIFT) | (adc3 << ADC3SHIFT))
 
-ADCf4xx::ADCf4xx(Pin pin) : ADC(pin) {
-    // Flag representing if the ADC has been configured yet
-    static bool halADCisInit = false;
+ADCf4xx::ADC_State_t ADCf4xx::adcArray[3];
+bool ADCf4xx::timerInit = false;
 
-    // "Rank" represents the order in which the channels are added
-    // Also represents the total number of added channels
-    static uint8_t rank = 1;
+ADCf4xx::ADCf4xx(Pin pin, ADCPeriph adcPeriph) : ADC(pin, adcPeriph) {
+    // Get adc state being updated
+    uint8_t adcNum                 = getADCNum();
+    ADCf4xx::ADC_State_t* adcState = &adcArray[adcNum];
```

Comment:
ok so i removed the adcf4xx:: but likeeee... how much better is ADC_State_t&, cause if i do that I gotta change all the -> to . :|

Reply:
Making it a reference means nobody can ever accidentally make it a null pointer, which is worth doing

Reply:
ok 😞

Reply:
Yo magee if I make it a reference, and also put it into the header so that I dont need to do "ADC_State_t& adcState = adcArray[adcNum];" in every method, then I need to have it passed in as a parameter the constructor. 

But for it to be a parameter in the constructor then I need to have the parameter in the constructor in the hpp file, but the constructor in the public section above the declaration of the struct, so it doesn't see the struct existing. Honestly idk what im doing w these in the hpp, I don't do stuff like this. Should I just like, make a second private section at the top of the header so its above the constructor, or what...

Reply:
Why would you pass it as a parameter to the constructor? Don't you figure out which one it should be based on  the result of `getADCNum()`?

Reply:
yes. So because of one of ur comments elsewhere, I moved the `ADC_State_t* adcState;` to the hpp so that I don't need to call and set it in every function. 
But if I try to change it to a reference (`ADC_State_t& adcState;`) in the hpp, it give the error 
"Constructor for 'core::io::ADCf4xx' must explicitly initialize the reference member 'adcState'" and tells me to add adcState to the parameters for the constructor. I'm not sure if there is another way to do it but that is what it tells me in the error

Reply:
oh wait, I think I got it to work. i didnt know you could just have whatever values you wanted in the member initialization list. I thought they had to be variables from the parameters... I added it and will push, and have u look at it again to make sure this is right